From 054a6cf3b4881c555e20410e83211cef8156ecc2 Mon Sep 17 00:00:00 2001
From: Robert S Peterson <rpeterso@redhat.com>
Date: Tue, 11 Nov 2014 18:10:26 -0500
Subject: [fs] GFS2: If we use up our block reservation, request more next time

Message-id: <6356d3a57a0a57b62320e93ae5c8f8d91ba130f6.1415729212.git.rpeterso@redhat.com>
Patchwork-id: 99829
O-Subject: [RHEL7.1 PATCH fs 3/3] GFS2: If we use up our block reservation, request more next time
Bugzilla: 1142238
RH-Acked-by: Abhijith Das <adas@redhat.com>
RH-Acked-by: Steven Whitehouse <swhiteho@redhat.com>

If we run out of blocks for a given multi-block allocation, we obviously
did not reserve enough. We should reserve more blocks for the next
reservation to reduce fragmentation. This patch increases the size hint
for reservations when they run out.

rhbz#1142238
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/gfs2/rgrp.c b/fs/gfs2/rgrp.c
index fb4041a..abeb9d7 100644
--- a/fs/gfs2/rgrp.c
+++ b/fs/gfs2/rgrp.c
@@ -2203,6 +2203,9 @@ static void gfs2_adjust_reservation(struct gfs2_inode *ip,
    trace_gfs2_rs(rs, TRACE_RS_CLAIM);
    if (rs->rs_free && !ret)
     goto out;
+   /* We used up our block reservation, so we should
+      reserve more blocks next time. */
+   atomic_add(RGRP_RSRV_ADDBLKS, &rs->rs_sizehint);
   }
   __rs_deltree(rs);
  }
diff --git a/fs/gfs2/rgrp.h b/fs/gfs2/rgrp.h
index 0ce3776..af87509 100644
--- a/fs/gfs2/rgrp.h
+++ b/fs/gfs2/rgrp.h
@@ -20,6 +20,7 @@
  */
 #define RGRP_RSRV_MINBYTES 8
 #define RGRP_RSRV_MINBLKS ((u32)(RGRP_RSRV_MINBYTES * GFS2_NBBY))
+#define RGRP_RSRV_ADDBLKS 64
 
 struct gfs2_rgrpd;
 struct gfs2_sbd;
-- 
1.7.1